flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/utils/MutableIOMetrics.java

```diff
@@ -79,6 +85,7 @@ public void addIOMetrics(AccessExecution attempt, @Nullable MetricFetcher fetche
 					this.numBytesOut += Long.valueOf(metrics.getMetric(MetricNames.IO_NUM_BYTES_OUT, "0"));
 					this.numRecordsIn += Long.valueOf(metrics.getMetric(MetricNames.IO_NUM_RECORDS_IN, "0"));
 					this.numRecordsOut += Long.valueOf(metrics.getMetric(MetricNames.IO_NUM_RECORDS_OUT, "0"));
+					this.metricsFetched = true;
```

Comment:
Note that this does not necessarily mean that the num/bytes IO metrics have been added yet; it only means that _some_ metric for the given subtask has already arrived and was inserted into the store.

Reply:
luckily this is only the cause because of a bug in this class. Access to the metric store must be guarded by a synchronization block on the metric store, which would then guarantee that the update from a taskmanager is applied in a all-or-nothing fashion.

Reply:
Unfortunately, if #4472 goes through that neat property of the synchronization goes away.